Lib/typing.py

```diff
@@ -308,19 +308,67 @@ def _unpack_args(args):
             newargs.append(arg)
     return newargs
 
-def _deduplicate(params):
+def _deduplicate(params, *, unhashable_fallback=True):
     # Weed out strict duplicates, preserving the first of each occurrence.
-    all_params = set(params)
-    if len(all_params) < len(params):
-        new_params = []
-        for t in params:
-            if t in all_params:
-                new_params.append(t)
-                all_params.remove(t)
-        params = new_params
-        assert not all_params, all_params
-    return params
-
+    try:
+        all_params = set(params)
+    except TypeError:
+        if not unhashable_fallback:
+            raise
+
+        # Happens for cases like `Annotated[dict, {'x': IntValidator()}]`
+        hasable, unhashable = _deduplicate_unhashable(params)
+        return hasable + unhashable
+
+    if len(all_params) == len(params):  # fast case
+        return params
+    new_params = []
+    for t in params:
```

Comment:
Can't this be done faster with `list(dict.fromkeys(params))`?

Reply:
I don't think we even need to convert it to a `list` before returning it: if you look at the callsites, this function just needs to return a deduped iterable that retains the order of the input. No tests fail if I apply this diff to @sobolevn's PR branch:

```diff
--- a/Lib/typing.py
+++ b/Lib/typing.py
@@ -311,7 +311,7 @@ def _unpack_args(args):
 def _deduplicate(params, *, unhashable_fallback=True):
     # Weed out strict duplicates, preserving the first of each occurrence.
     try:
-        all_params = set(params)
+        deduped = dict.fromkeys(params)
     except TypeError:
         if not unhashable_fallback:
             raise
@@ -319,15 +319,7 @@ def _deduplicate(params, *, unhashable_fallback=True):
         # Happens for cases like `Annotated[dict, {'x': IntValidator()}]`
         return _deduplicate_unhashable(params)

-    if len(all_params) == len(params):  # fast case
-        return params
-    new_params = []
-    for t in params:
-        if t in all_params:
-            new_params.append(t)
-            all_params.remove(t)
-    assert not all_params, all_params
-    return new_params
+    return deduped
```